File: mbflow-ui/src/types/workflow.ts
export type WorkflowStatus = "draft" | "active" | "inactive" | "archived";

export interface Position {
  x: number;
  y: number;
}

export interface Node {
  id: string;
  name: string;
  type: string;
  description?: string;
  config: Record<string, any>;
  position?: Position;
  metadata?: Record<string, any>;
}

export interface Edge {
  id: string;
  from: string;
  to: string;
  condition?: string;
  metadata?: Record<string, any>;
}

export interface Workflow {
  id: string;
  name: string;
  description?: string;
  version: number;
  status: WorkflowStatus;
  tags?: string[];
  nodes: Node[];
  edges: Edge[];
  variables?: Record<string, any>;
  metadata?: Record<string, any>;
  created_at: string;
  updated_at: string;
}

export interface WorkflowCreateRequest {
  name: string;
  description?: string;
  variables?: Record<string, any>;
  metadata?: Record<string, any>;
}

export interface WorkflowUpdateRequest {
  name?: string;
  description?: string;
  variables?: Record<string, any>;
  metadata?: Record<string, any>;
  status?: WorkflowStatus;
}

File: backend/sdk/go/models/workflow.go
package models

import "time"

// WorkflowStatus represents the lifecycle state of a workflow.
type WorkflowStatus string

const (
	WorkflowStatusDraft    WorkflowStatus = "draft"
	WorkflowStatusActive   WorkflowStatus = "active"
	WorkflowStatusInactive WorkflowStatus = "inactive"
	WorkflowStatusArchived WorkflowStatus = "archived"
)

// Workflow represents a complete workflow definition with its DAG structure.
type Workflow struct {
	ID          string         `json:"id"`
	Name        string         `json:"name"`
	Description string         `json:"description,omitempty"`
	Version     int            `json:"version"`
	Status      WorkflowStatus `json:"status"`
	Tags        []string       `json:"tags,omitempty"`
	Nodes       []*Node        `json:"nodes"`
	Edges       []*Edge        `json:"edges"`
	Variables   map[string]any `json:"variables,omitempty"`
	Metadata    map[string]any `json:"metadata,omitempty"`
	CreatedBy   string         `json:"created_by,omitempty"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
}

// Node represents a single node in the workflow DAG.
type Node struct {
	ID          string         `json:"id"`
	Name        string         `json:"name"`
	Type        string         `json:"type"`
	Description string         `json:"description,omitempty"`
	Config      map[string]any `json:"config"`
	Position    *Position      `json:"position,omitempty"`
	Metadata    map[string]any `json:"metadata,omitempty"`
}

// Position represents the visual position of a node in the editor.
type Position struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

// Edge represents a directed connection between two nodes in the DAG.
type Edge struct {
	ID           string         `json:"id"`
	From         string         `json:"from"`
	To           string         `json:"to"`
	SourceHandle string         `json:"source_handle,omitempty"`
	Condition    string         `json:"condition,omitempty"`
	Loop         *LoopConfig    `json:"loop,omitempty"`
	Metadata     map[string]any `json:"metadata,omitempty"`
}

// LoopConfig configures a loop edge that allows controlled re-execution of a wave range.
type LoopConfig struct {
	MaxIterations int `json:"max_iterations"`
}
